hpe3parclient/client.py

```diff
@@ -968,32 +996,44 @@ def copyVolume(self, src_name, dest_name, dest_cpg, optional=None):
         # has to be taken care by caller side
         if optional:
             if self.primera_supported:
-                # For primera array there is no compression and tdvv parameters
-                # removing tdvv and replacing compression with reduce field
-                if 'tdvv' in optional:
-                    optional.pop('tdvv')
-
-                if optional.get('tpvv') is True \
-                        and optional.get('compression') is True:
-                    optional['reduce'] = True
+                for key in ['tpvv', 'compression', 'tdvv']:
+                    option = optional.get(key)
+                    if option and option not in [True, False]:
+                        # raising exception for junk compression input
+                        ex_desc = "39 - invalid input: wrong type for key " \
+                            "[%s]. Valid values are [True, False]" % key
+                        raise exceptions.HTTPBadRequest(ex_desc)
+
+                if optional.get('compression') is True:
+                    combination = ['tdvv', 'compression']
+                    len_diff = len(set(combination) - set(optional.keys()))
+                    msg = "invalid input: For compressed and deduplicated "\
+                          "volumes both 'compression' and " \
+                          "'tdvv' must be specified as true"
+                    if len_diff == 1:
+                        raise exceptions.HTTPBadRequest(msg)
+                    if optional.get('tdvv') is True \
+                            and optional.get('compression') is True:
+                        optional['reduce'] = True
+
+                    if optional.get('tdvv') is False \
+                            and optional.get('compression') is True:
+                        raise exceptions.HTTPBadRequest(msg)
+                else:
+                    msg = "invalid input: For compressed and deduplicated "\
+                          "volumes both 'compression' and "\
+                          "'tdvv' must be specified as true"
+                    if optional.get('tdvv') is False \
+                            and optional.get('compression') is False:
```

Comment:
@bhagyashree-sarawate Won't compression be always false in the else block?

Reply:
@sheharai4  yes, in else part compression will be false but keeping the conditions same as before for the clear understanding, what should be done when it comes along with tdvv

Reply:
Else part can have optional.get('compression') == None if the key is not present.. so explicit `optional.get('compression')` is False is required.